runtime/runtime/src/pipelining.rs

```diff
@@ -0,0 +1,182 @@
+use near_parameters::RuntimeConfig;
+use near_primitives::account::Account;
+use near_primitives::action::Action;
+use near_primitives::config::ViewConfig;
+use near_primitives::hash::CryptoHash;
+use near_primitives::receipt::{Receipt, ReceiptEnum};
+use near_primitives::types::AccountId;
+use near_vm_runner::logic::GasCounter;
+use near_vm_runner::{ContractRuntimeCache, PreparedContract};
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::{Arc, Condvar, Mutex};
+
+use crate::ext::{ContractStorage, RuntimeContractExt};
+
+pub(crate) struct ReceiptPreparationPipeline {
+    /// Mapping from a Receipt's ID to a parallel "task" to prepare the receipt's data.
+    ///
+    /// The task itself may be run in the current thread, a separate thread or forced in any other
+    /// way.
+    map: BTreeMap<PrepareTaskKey, Arc<PrepareTask>>,
+
+    /// List of Receipt receiver IDs that must not be prepared for this chunk.
+    ///
+    /// This solves an issue wherein the pipelining implementation only has access to the committed
+    /// storage (read: data as a result of applying the previous chunk,) and not the state that has
+    /// been built up as a result of processing the current chunk. One notable thing that may have
+    /// occurred there is a contract deployment. Once that happens, we can no longer get the
+    /// "current" contract code for the account.
+    ///
+    /// However, even if we had access to the transaction of the current chunk and were able to
+    /// access the new code, there's a risk of a race between when the deployment is executed
+    /// and when a parallel preparation may occur, leading back to needing to hold prefetching of
+    /// that account's contracts until the deployment is executed.
+    ///
+    /// As deployments are a relatively rare event, it is probably just fine to entirely disable
+    /// pipelining for the account in question for that particular block. This field implements
+    /// exactly that.
+    ///
+    /// In the future, however, it may make sense to either move the responsibility of executing
+    /// deployment actions to this pipelining thingy OR, even better, modify the protocol such that
+    /// contract deployments in block N only take effect in the block N+1 as that, among other
+    /// things, would give the runtime more time to compile the contract.
+    block_accounts: BTreeSet<AccountId>,
+
+    /// The Runtime config for these pipelining  requests.
+    config: Arc<RuntimeConfig>,
+
+    /// The contract cache.
+    contract_cache: Option<Box<dyn ContractRuntimeCache>>,
+
+    /// The chain ID being processed.
+    chain_id: String,
+
+    /// Protocol version for this chunk.
+    protocol_version: u32,
+
+    /// Storage.
+    storage: Arc<dyn near_store::TrieStorage>,
```

Comment:
Question to the storage folks: is this the right way to get access to the contracts here? If not, what would be the right way? I feel like this code needn't be concerned about recording and such, because we already handle it separately through prefetching. The access mechanism has to be thread-safe.

To give some more background, the implementation here is meant to access the storage for the previous block than the one currently processed, largely because sharing `TrieUpdate` across threads is not really easy to think about. So the idea is to directly access RocksDB (or another implementation of storage), which would have changes from the previous block/chunk already committed.